File: tests/functionPointers.cc



void print_concat(char* a, char* b, int c)
{
  printf("Join 3: %s %s %i\n", a, b, c);
}

void (*custom_print)(char* a, char* b, int c) = print_concat;


void (*catch_and_return(void (*pf)(char* aarg, char* barg, int* ccarg), const char *name_one, const char* name_two, int* number))(char* a, char * b, int c)
{
  static int value = 42;
  pf((char*)"hello", (char*)"my", &value);
  return custom_print;
}


void pf(char* aarg, char* barg, int* ccarg) {

}

void test()
{
  int d = 5;
  (catch_and_return(pf,"b", "c", &d));
}


#include <build/cbind_functionPointers.h>

File: tests/functionPointers.js

var cBind = require('../build/Debug/cbind_core.node');

var functionPointers = cBind.functionPointers;

var mocha = require('mocha');

var chai = require('chai');
var assert = chai.assert;
chai.should();

describe('function pointers', function() {

  it('should call complex definition', function() {

    var innerCalled = false;
    var customPrintCalled = false;

    functionPointers.custom_print = function(a, b, c) {
      a.should.equal('FINAL');
      b.should.equal('TEST');
      c.should.equal(44);
      customPrintCalled = true;
    };

    functionPointers.catch_and_return(function(a, b, c) {

      a.should.equal('hello');
      b.should.equal('my');
      cBind.derefInt(c).should.equal(42);
      innerCalled = true;

    }, "str1", "str2", cBind.createInt(30))("FINAL", "TEST", 44);

    innerCalled.should.equal(true);
    customPrintCalled.should.equal(true);

  });

  it('should get the same object after setting to the variable', function() {

    var f = function(a, b, c) {};

    functionPointers.custom_print = f;

    functionPointers.custom_print.should.equal(f);



  });
});